File: frontend/serveo-admin/src/core/errors/api-problem-details.ts
export interface ApiProblemDetailsBase {
  type?: string;
  title?: string;
  status: number;
  detail?: string;
  instance?: string;
  traceId?: string;
}

export interface ApiProblemError {
  code: string;
  message: string;
  type: string;
  target?: string | null;
}

export interface ValidationProblemDetails extends ApiProblemDetailsBase {
  status: 400;
  errors: Record<string, string[]>;
}

export interface ErrorProblemDetails extends ApiProblemDetailsBase {
  status: 401 | 403 | 404 | 409 | 422 | 500;
  errors: ApiProblemError[];
}

export type ApiProblemDetails = ErrorProblemDetails | ValidationProblemDetails;

File: backend/Serveo.WebApi/Common/ApiProblemDetails.cs
using Microsoft.AspNetCore.Mvc;
using System.Text.Json.Serialization;

namespace Serveo.WebApi.Common
{
    // Model response cho API error
    public sealed class ApiProblemDetails : ProblemDetails
    {
        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
        public string? TraceId { get; init; }

        public List<ApiProblemError> Errors { get; init; } = [];
    }

    public sealed class ApiProblemError
    {
        public string Code { get; init; } = default!;
        public string Message { get; init; } = default!;

        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
        public string? Type { get; init; }


        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
        public string? Target { get; init; }
    }
}
